barretenberg/cpp/src/barretenberg/stdlib/hash/poseidon2/poseidon2_permutation_classic.cpp

```diff
@@ -0,0 +1,310 @@
+#include "poseidon2_permutation_classic.hpp"
+
+#include "barretenberg/proof_system/arithmetization/gate_data.hpp"
+#include "barretenberg/proof_system/circuit_builder/goblin_ultra_circuit_builder.hpp"
+#include "barretenberg/stdlib/primitives/circuit_builders/circuit_builders_fwd.hpp"
+
+namespace bb::stdlib {
+
+template <typename Params, typename Builder>
+void Poseidon2PermutationClassic<Params, Builder>::add_round_constants(
+    State& input, const typename Poseidon2PermutationClassic<Params, Builder>::RoundConstants& rc)
+{
+    for (size_t i = 0; i < t; ++i) {
+        input[i] += rc[i];
+    }
+}
+
+template <typename Params, typename Builder> void Poseidon2PermutationClassic<Params, Builder>::apply_sbox(State& input)
+{
+    for (auto& in : input) {
+        apply_single_sbox(in);
+    }
+}
+
+template <typename Params, typename Builder>
+void Poseidon2PermutationClassic<Params, Builder>::apply_single_sbox(field_t<Builder>& input)
+{
+    // hardcoded assumption that d = 5. should fix this or not make d configurable
+    auto xx = input.sqr();
+    auto xxxx = xx.sqr();
+    input *= xxxx;
+}
+
+template <typename Params, typename Builder>
+void Poseidon2PermutationClassic<Params, Builder>::matrix_multiplication_internal(State& input)
+{
+    // for t = 4
+    auto sum = input[0];
+    for (size_t i = 1; i < t; ++i) {
+        sum += input[i];
+    }
+    for (size_t i = 0; i < t; ++i) {
+        input[i] *= Params::internal_matrix_diagonal[i]; // internal_matrix_diagonal[i];
+        input[i] += sum;
+    }
+}
+
+template <typename Params, typename Builder>
+typename Poseidon2PermutationClassic<Params, Builder>::State Poseidon2PermutationClassic<Params, Builder>::
+    goblin_permutation(Builder* builder, const typename Poseidon2PermutationClassic<Params, Builder>::State& input)
+{
+    // deep copy
+    State current_state(input);
+    NativeState current_native_state;
+    for (size_t i = 0; i < t; ++i) {
+        current_native_state[i] = current_state[i].get_value();
+    }
+
+    // Apply 1st linear layer
+    NativePermutation::matrix_multiplication_external(current_native_state);
+    initial_external_matrix_multiplication(builder, current_state);
+
+    // First set of external rounds
+    constexpr size_t rounds_f_beginning = rounds_f / 2;
+    for (size_t i = 0; i < rounds_f_beginning; ++i) {
+        poseidon2_external_gate_<FF> in{ current_state[0].witness_index,
+                                         current_state[1].witness_index,
+                                         current_state[2].witness_index,
+                                         current_state[3].witness_index,
+                                         i };
+        ((GoblinUltraCircuitBuilder*)builder)->create_poseidon2_external_gate(in);
```

Comment:
doesn't seem very clean to me. I think just restrict this function to IsGoblinBuilder<Builder> Builder in the template parameter